graylog2-server/src/main/java/org/graylog/plugins/views/search/engine/PositionTrackingQuery.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.plugins.views.search.engine;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class PositionTrackingQuery {
+    private final List<QueryFragment> fragments;
+
+    public PositionTrackingQuery(List<QueryFragment> fragments) {
+        this.fragments = fragments;
+    }
+
+    public static PositionTrackingQuery of(String queryString) {
+        return new PositionTrackingQuery(Collections.singletonList(
+                new QueryFragment(queryString, queryString, 1, 0, queryString.length())
+        ));
+    }
+
+    public List<QueryFragment> getFragments() {
+        return fragments;
+    }
+
+    public String getOriginalQuery() {
+        return concatFragments(QueryFragment::getOriginalContent);
+    }
+
+    public String getInterpolatedQuery() {
+        return concatFragments(QueryFragment::getInterpolatedContent);
+    }
+
+    private String concatFragments(Function<QueryFragment, String> supplier) {
+        int line = 1;
+        final StringBuilder stringBuilder = new StringBuilder();
+        for (QueryFragment fragment : fragments) {
+            if (line != fragment.getLine()) {
+                stringBuilder.append("\n");
+                line = fragment.getLine();
+            }
+            stringBuilder.append(supplier.apply(fragment));
+        }
+        return stringBuilder.toString();
```

Comment:
I am not sure if I understand this correctly, but is there an implicit assumption that the fragments in the list have growing/ordered line numbers? If it is true, I'd say this assumption is dangerous (client code may stop following it at any time) and I would rather go for more precise collection for storing fragments, i.e. `List<List<QueryFragment>> linesWithFragments`... or sorting fragments with a comparator based on line?

Reply:
The fragments are ordered, but you are right that this happens outside of this class. Adding an explicit sort just to be sure.